File: Level3/가장 긴 팰린드롬.cpp
#include <string>

using namespace std;

int solution(string s) {

	int max_len = 1;
	int len, begin, end;

	for (int i = 0; i < s.length() - 1; i++) {
		// case 1
		if (s[i] == s[i + 1]) {
			len = 2, begin = i - 1, end = i + 2;
			while (begin >= 0 && end < s.length()) {
				if (s[begin--] != s[end++]) break;
				len += 2;
			}
			max_len = max(max_len, len);
		}
		// case 2
		if (i > 0 && s[i - 1] == s[i + 1]) {
			len = 3, begin = i - 2, end = i + 2;
			while (begin >= 0 && end < s.length()) {
				if (s[begin--] != s[end++]) break;
				len += 2;
			}
			max_len = max(max_len, len);
		}
	}

	return max_len;
}

File: programmers-coding-test/Python/Level3/가장 긴 팰린드롬.py
def solution(s):
    max_length = 1
    for i in range(len(s) - 1):
        if s[i] == s[i + 1]:
            begin, end, length = i - 1, i + 2, 2
            while begin in range(len(s)) and end in range(len(s)):
                if s[begin] != s[end]:
                    break
                begin, end, length = begin - 1, end + 1, length + 2
            max_length = max(max_length, length)
        if i > 0 and s[i - 1] == s[i + 1]:
            begin, end, length = i - 2, i + 2, 3
            while begin in range(len(s)) and end in range(len(s)):
                if s[begin] != s[end]:
                    break
                begin, end, length = begin - 1, end + 1, length + 2
            max_length = max(max_length, length)
    return max_length
